desc/batching.py

```diff
@@ -58,145 +69,213 @@ def _unchunk(x):
 def _chunk(x, chunk_size=None):
     # chunk_size=None -> add just a dummy chunk dimension,
     #  same as np.expand_dims(x, 0)
-    if x.ndim == 0:
-        raise ValueError("x cannot be chunked as it has 0 dimensions.")
+    errorif(x.ndim == 0, msg="x cannot be chunked as it has 0 dimensions.")
     n = x.shape[0]
-    if chunk_size is None:
-        chunk_size = n
+    chunk_size = setdefault(chunk_size, n)
 
     n_chunks, residual = divmod(n, chunk_size)
-    if residual != 0:
-        raise ValueError(
-            "The first dimension of x must be divisible by chunk_size."
-            + f"\n        Got x.shape={x.shape} but chunk_size={chunk_size}."
-        )
+    errorif(
+        residual != 0,
+        msg="The first dimension of x must be divisible by chunk_size.\n"
+        + f"Got x.shape={x.shape} but chunk_size={chunk_size}.",
+    )
     return x.reshape((n_chunks, chunk_size) + x.shape[1:])
 
 
-####
+_tree_zeros_like = partial(
+    jax.tree_util.tree_map, lambda x: jnp.zeros(x.shape, dtype=x.dtype)
+)
 
-# The following section of this code is derived from the NetKet project
-# https://github.com/netket/netket/blob/9881c9fb217a2ac4dc9274a054bf6e6a2993c519/
-# netket/jax/_scanmap.py
 
+def _identity(y):
+    """Returns the input."""
+    return y
 
-def _scan_append(f, x):
-    """Evaluate f element by element in x while appending the results.
 
-    Parameters
-    ----------
-    f: a function that takes elements of the leading dimension of x
-    x: a pytree where each leaf array has the same leading dimension
+def _multimap(f, *args):
+    try:
+        return tuple(f(*a) for a in zip(*args))
+    except TypeError:
+        return f(*args)
 
-    Returns
-    -------
-    a (pytree of) array(s) with leading dimension same as x,
-    containing the evaluation of f at each element in x
-    """
-    carry_init = True
+
+def _scan_append_reduce(
+    f,
+    x,
+    reduction=None,
+    zero_fun=_tree_zeros_like,
+):
+    """Evaluate f element-wise in x while appending and/or reducing the results."""
+    x0 = jax.tree_util.tree_map(lambda x: x[0], x)
+
+    # special code path if there is only one element
+    # to avoid having to rely on xla/llvm to optimize the overhead away
+    if jax.tree_util.tree_leaves(x)[0].shape[0] == 1:
+        return _multimap(
+            lambda c, x: jnp.expand_dims(x, 0) if c else x,
+            reduction is None,
+            f(x0),
+        )
+
+    # the original idea was to use pytrees, however for now just operate on the
+    # return value tuple
```

Comment:
i kept this clunky pattern in case they follow through on this comment, so we can port it easier